lighthouse-core/computed/js-bundles.js

```diff
@@ -0,0 +1,129 @@
+  // If the map + contents don't line up, return a result that
+  // denotes nothing is mapped.
+  const failureResult = {files: {}, unmappedBytes, totalBytes};
+
+  // @ts-ignore: This function is added in SDK.js
+  map.computeLastGeneratedColumns();
+
+  for (const mapping of map.mappings()) {
+    const source = mapping.sourceURL;
+    const lineNum = mapping.lineNumber;
+    const colNum = mapping.columnNumber;
+    // @ts-ignore: `lastColumnNumber` is not on types yet.
+    const lastColNum = /** @type {number=} */ (mapping.lastColumnNumber);
+
+    // Webpack sometimes emits null mappings.
+    // https://github.com/mozilla/source-map/pull/303
+    if (!source) continue;
+
+    // Lines and columns are zero-based indices. Visually, lines are shown as a 1-based index.
+
+    const line = lines[lineNum];
+    if (line === null) {
+      log.error('JSBundles', `${map.url()} mapping for line out of bounds: ${lineNum + 1}`);
+      return failureResult;
+    }
+
+    if (colNum > line.length) {
+      // eslint-disable-next-line max-len
+      log.error('JSBundles', `${map.url()} mapping for column out of bounds: ${lineNum + 1}:${colNum}`);
+      return failureResult;
+    }
+
+    let mappingLength = 0;
+    if (lastColNum !== undefined) {
+      if (lastColNum > line.length) {
+        // eslint-disable-next-line max-len
+        log.error('JSBundles', `${map.url()} mapping for last column out of bounds: ${lineNum + 1}:${lastColNum}`);
+        return failureResult;
+      }
+      mappingLength = lastColNum - colNum;
+    } else {
+      // TODO Buffer.byteLength?
+      // Add +1 to account for the newline.
+      mappingLength = line.length - colNum + 1;
+    }
+    files[source] = (files[source] || 0) + mappingLength;
+    unmappedBytes -= mappingLength;
+  }
+
+  return {
+    files,
+    unmappedBytes,
+    totalBytes,
+  };
+}
+
+class JSBundles {
+  /**
+   * @param {Pick<LH.Artifacts, 'SourceMaps'|'ScriptElements'|'devtoolsLogs'>} artifacts
+   * @param {LH.Audit.Context} context
+   */
+  static async compute_(artifacts, context) {
+    const {SourceMaps, ScriptElements} = artifacts;
+    const devtoolsLog = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
+    const networkRecords = await NetworkRecords.request(devtoolsLog, context);
+
+    /** @type {LH.Artifacts.Bundle[]} */
+    const bundles = [];
+
+    // Collate map, script, and network record.
+    for (const SourceMap of SourceMaps) {
+      if (!SourceMap.map) continue;
+      const {scriptUrl, map: rawMap} = SourceMap;
+
+      if (!rawMap.mappings) continue;
+
+      const scriptElement = ScriptElements.find(s => s.src === scriptUrl);
+      const networkRecord = networkRecords.find(r => r.url === scriptUrl);
+      if (!scriptElement) continue;
+
+      const compiledUrl = SourceMap.scriptUrl || 'compiled.js';
+      const mapUrl = SourceMap.sourceMapUrl || 'compiled.js.map';
+      // @ts-ignore: CDT has some funny ideas about what properties of a source map are required.
```

Comment:
this one concerns me a little, is it mostly that they require `key: T|undefined` but ours is `key?: T` or something? would nice to make these equivalent to catch any type bugs if the CDT definition ever changes

Reply:
CDT expects properties to be explicitly undefined, instead of implicitly undefined. Not a result of a specific design decision, just a result of how CDT marks its types (see `SourceMapV3` ctor).

fixed.

Reply:
ugh, the recursive nature of `sections` breaks this. Can't reasonably work around this. Should go back to a ts-ignore.